include/aspect/introspection.h

```diff
@@ -29,10 +29,265 @@
+   */
+  template <int dim>
+  class IntrospectionBase
+  {
+    public:
+      /**
+         * Construct an empty object.
+         */
+      IntrospectionBase() {}
+
+      /**
+         * Construct object from a list of variables (identical to calling initialize()).
+         */
+      IntrospectionBase(const std::vector<Variable<dim> > &variables)
+      {
+        initialize(variables);
+      }
+
+      /**
+       * Fill this object with the given list of @p variables.
+       */
+      void initialize(const std::vector<Variable<dim> > &variables)
+      {
+        variables_ = variables;
+        name_map.clear();
+
+        unsigned int component_index = 0;
+        unsigned int block_index = 0;
+
+        for (unsigned int i=0; i<variables_.size(); ++i)
+          {
+            variables_[i].initialize(component_index, block_index, i);
+            component_index+= variables_[i].n_components();
+            block_index += variables_[i].n_blocks;
+
+            Assert(name_map.find(variables_[i].name)==name_map.end(),
+                   ExcMessage("Can not add two variables with the same name."));
+            name_map.insert(std::make_pair(variables_[i].name, &variables_[i]));
+          }
+        Assert(variables_.back().n_blocks != 0
+               || variables_.back().n_components() == 0
+               , ExcMessage("The last variable needs to have >0 blocks."));
+
+        n_components_ = component_index;
+        n_blocks_ = block_index;
+
+        fes.resize(variables_.size());
+        multiplicities.resize(variables_.size());
+        for (unsigned int i=0; i<variables_.size(); ++i)
+          {
+            fes[i] = &*(variables_[i].fe);
+            multiplicities[i] = variables_[i].multiplicity;
+            variables_[i].component_mask=ComponentMask(n_components_, false);
+            for (unsigned int c=0; c<variables_[i].n_components(); ++c)
+              variables_[i].component_mask.set(c+variables_[i].first_component_index, true);
+          }
+
+        components_to_blocks_.clear();
+        for (unsigned int i=0; i<variables_.size(); ++i)
+          {
+            for (unsigned int c=0; c<variables_[i].n_components(); ++c)
+              components_to_blocks_.push_back(variables_[i].block_index
+                                              + ((variables_[i].n_blocks>1)?c:0));
+          }
+      }
+
+      /**
+       * Return the variable with name @p name. Throws an exception if this variable
+       * does not exist.
+       */
+      const Variable<dim> &variable(const std::string &name) const
+      {
+        typename std::map<std::string, Variable<dim>*>::const_iterator it
+          = name_map.find(name);
+        Assert(it != name_map.end(), ExcMessage("Variable '" + name + "' not found!"));
+        return *(it->second);
+      }
+
+      /**
+       * Return the list of all variables.
+       */
+      const std::vector<Variable<dim> > &variables() const
```

Comment:
And maybe we could give this function a more specific name too, something like `get_list_of_variables`, or `get_all_variables`?